gk/main.c

```diff
@@ -2211,8 +2211,13 @@ gk_proc(void *arg)
 			 * an entry has been freed from the flow table.
 			 */
 			entry_idx = (entry_idx + 1) % gk_conf->flow_ht_size;
-			if (gk_flow_tbl_entry_scan(entry_idx, instance))
-				instance->has_insertion_failed = false;
+			if (gk_flow_tbl_entry_scan(entry_idx, instance)) {
+				if (++instance->num_scan_del >=
+						gk_conf->scan_del_thresh) {
+					instance->num_scan_del = 0;
+					instance->has_insertion_failed = false;
```

Comment:
This code updates `instance->num_scan_del` even when `instance->has_insertion_failed` is false. So when `instance->has_insertion_failed` is set true, `instance->num_scan_del` can have any value.

A way to address this issue is to do the following:
1. Drop field `instance->has_insertion_failed`, and in its place use test `instance->num_scan_del == 0`;
2. Instead of setting `instance->has_insertion_failed` true, use `instance->num_scan_del == gk_conf->scan_del_thresh`;
3. Only decrement `instance->num_scan_del` if it is not zero.

Reply:
2. Shouldn't it be ``instance->num_scan_del = gk_conf->scan_del_thresh`` instead of ``instance->num_scan_del == gk_conf->scan_del_thresh``?

Reply:
You are right. Thank you.

Reply:
1. To test there is an insertion failure, shouldn't we test ``instance->num_scan_del > 0``? ``instance->num_scan_del == 0`` means that there is no insertion failure, it's free to insert new flow entries.

Reply:
You are right again.